doc/developer/design/20230714_show_privileges.md

```diff
@@ -0,0 +1,202 @@
+# SHOW PRIVILEGES
+
+Associated:
+
+- [GitHub issue](https://github.com/MaterializeInc/materialize/issues/20452)
+- [RBAC design doc](20230216_role_based_access_control.md)
+- [System privilege design doc](20230531_system_privileges.md)
+- [Default privilege design doc](20230602_default_privileges.md)
+
+## Context
+
+Role based access control comprises a set of features that allow customers to control the
+privileges of individual users. Privileges can be granted to a user which allows them to take
+certain actions in a Materialize deployment. All objects have an owner, and certain actions are
+limited to only object owners. Roles can be members of other roles and inherit all of their
+privileges. Privileges, owners, and role membership are all stored within the database catalog.
+
+Currently, it's difficult for a user to discover RBAC related information.
+
+## Goals
+
+- Make it easy for users to discover the set of privileges that they have.
+- Make it easy for users to discover the owner of objects.
+- Make it easy for users to discover the roles that they are a member of.
+
+## Overview
+
+We will add the following SQL commands:
+
+- `SHOW PRIVILEGES`: shows the privileges of the current user.
+- `SHOW DEFAULT PRIVILEGES`: shows default privileges.
+- `SHOW ROLE MEMBERSHIP`
+
+We will also add an `owner` column to the existing show commands.
+
+## Detailed description
+
+- We will add an internal unmaterializable function, `mz_internal.mz_role_membership() -> text[]`,
+  that returns an array of all the role IDs that the session's current role is a member of.
```

Comment:
Why do we need this function? Can't we just look up `current_role` (or whatever it's called) in the `mz_role_memberships` table (or whatever it's called)?

Reply:
`mz_role_members` shows direct membership, so if I had the following grants:

```SQL
GRANT r2 TO r1;
GRANT r3 TO r2;
```

Then I'd see the following in `mz_role_members`:
```
role_id | member
--------+--------
r2      | r1
r3      | r2
```

However `r1` really has the privileges of both `r2` and `r3`. We'd have to look this up using a recursive CTE, and I don't think we can use recursive CTEs to implement `SHOW` commands. A `SHOW` command can be nested arbitrarily deep in a query and expands into some `SELECT`, and I don't think we can't expand a CTE definition in the middle of a query. Though maybe we can because it's not expanded until planning, I'd have to look into that.